### ISSUE TITLE
False positive BlockAndLocalOverlapViolation with type annotation

### ISSUE THREAD

**Issue by AlwxSin** [wemake-services/wemake-python-styleguide#987]
# Bug report

## What's wrong
With given code
```python
import asyncpg as apg


async def db_operation(db: apg.pool.Pool):
    """Operation with db."""
    connection: apg.Connection
    async with db.acquire() as connection:
        return await connection.fetch(
            """
            SELECT * FROM table
            """
        )
```
`BlockAndLocalOverlapViolation` raises with 
```
     7:   5 WPS440 Found block variables overlap: connection [wemake-python-styleguide]
  async with db.acquire() as connection:
  ^
```

Without line ```connection: apg.Connection``` everything is ok.

## How is that should be
This is valid python annotation and `BlockAndLocalOverlapViolation` violation should not be raised here.

**Comment by sobolevn** [comment]
Yeap, this is a bug. @AlwxSin do you want to fix it?

**Comment by AlwxSin** [comment]
@sobolevn Yep, I'll do it. Assign on me, please.

**Comment by AlwxSin** [comment]
@sobolevn well, I've got a problem. I added one more `_scope_predicate` to `BlockVariableVisitor` which returns `True` if node is `AnnAssign` and there is no `value` in it. But the problem is in code to test
```python
def context():
    conn: Connection
    with db.get_conn() as conn:
        ...
```
`with` block has no parent, so [here](https://github.com/wemake-services/wemake-python-styleguide/blob/master/wemake_python_styleguide/visitors/ast/blocks.py#L155) node is `None` and obviously node is not `AnnAssign`.

`BlockVariableVisitor` is fragile and complex and I'm not sure how to fix it.

**Comment by sobolevn** [comment]
@AlwxSin you are interested in this method: https://github.com/wemake-services/wemake-python-styleguide/blob/master/wemake_python_styleguide/visitors/ast/blocks.py#L161

Here's how the flow works:
1. We start with `visit_locals`
2. Then we fall into `_scope` with `node: ast.AnnAssign`
3. Then we check `ignored_name` in `_naming_predicates`: https://github.com/wemake-services/wemake-python-styleguide/blob/master/wemake_python_styleguide/visitors/ast/blocks.py#L194-L197

That's where you need to add a new check: if it is a `AnnAssign` without a value, than the name is ignored.

By the way, if you find this code complex an unreadable we should at least write some docs into it. Do you want to help with it? Basically, all the same things I wrote here.

**Comment by AlwxSin** [comment]
> if you find this code complex and unreadable
I just copied docstring which said that it fragile and complex 😏

**Comment by AlwxSin** [comment]
Done. My mistake was that I thought only about `scope_predicates`, not `naming_predicates`.
